cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/ArtifactCache.java

```diff
@@ -0,0 +1,167 @@
+ * <last-modified-timestamp>.jar
+ *
+ */
+public class ArtifactCache extends AbstractArtifactLocalizer {
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactCache.class);
+
+  @Inject
+  public ArtifactCache(CConfiguration cConf)
+    throws InstantiationException, IllegalAccessException {
+    super(cConf.get(Constants.ArtifactCache.LOCAL_DATA_DIR),
+          RetryStrategies.fromConfiguration(cConf, Constants.Service.ARTIFACT_CACHE + "."));
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tethering.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      RemoteAuthenticator authenticator = authClass.newInstance();
+      RemoteAuthenticator.setDefaultAuthenticator(authenticator);
+    }
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as
+   * well as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @param peer The name of the tethered peer
+   * @param remoteClient The remote client used to connect to appfabric on the peer
+   * @return The Local Location for this artifact
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getArtifact(ArtifactId artifactId, String peer, RemoteClient remoteClient) throws Exception {
+    return Retries.callWithRetries(() -> fetchArtifact(artifactId, peer, remoteClient), retryStrategy);
+  }
+
+  /**
+   * fetchArtifact attempts to connect to app fabric to download the given artifact. This method will throw
+   * {@linkRetryableException} in certain circumstances.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @param peer The name of the tethered peer
+   * @param remoteClient The remote client used to connect to appfabric on the peer
+   * @return The Local Location for this artifact
+   * @throws IOException If an unexpected error occurs while writing the artifact to the filesystem
+   * @throws ArtifactNotFoundException If the given artifact does not exist
+   */
+  private File fetchArtifact(ArtifactId artifactId, @Nullable String peer, RemoteClient remoteClient)
+    throws IOException, ArtifactNotFoundException {
+    File artifactDir = getArtifactDirLocation(artifactId, peer);
+    Long lastModifiedTimestamp = getCurrentLastModifiedTimestamp(artifactDir);
+    HttpURLConnection urlConn = openConnection(artifactId, remoteClient);
+
+    try {
+      if (lastModifiedTimestamp != null) {
+        LOG.debug("Found existing local version for {} with timestamp {}", artifactId, lastModifiedTimestamp);
+
+        ZonedDateTime lastModifiedDate = ZonedDateTime
```

Comment:
This part is somewhat repetitive across the cache and the localizer classes. Can the parent class provide a method like `fetchArtifact(URL url, RemoteClient remoteClient)` to perform the common logic? Children classes is only responsible for generating the URL and to provide a remote client to talk to the URL.

Reply:
fixed this by creating method `fetchArtifact(ArtifactId artifactId, RemoteClient remoteClient, File artifactDir)` in the parent class. The child classes just provide the remote client and the artifact directory